controller_manager/src/controller_manager.cpp

```diff
@@ -158,4 +882,58 @@ ControllerManager::cleanup() const
   return ret;
 }
 
+std::vector<ControllerSpec> &
+ControllerManager::RTControllerListWrapper::update_and_get_used_by_rt_list()
+{
+  used_by_realtime_controllers_index_ = updated_controllers_index_;
+  return controllers_lists_[used_by_realtime_controllers_index_];
+}
+
+std::vector<ControllerSpec> &
+ControllerManager::RTControllerListWrapper::get_unused_list(
+  const std::lock_guard<std::recursive_mutex> &)
+{
+  assert(controllers_lock_.try_lock());
+  controllers_lock_.unlock();
+  // Get the index to the outdated controller list
+  int free_controllers_list = get_other_list(updated_controllers_index_);
+
+  // Wait until the outdated controller list is not being used by the realtime thread
+  wait_until_rt_not_using(free_controllers_list);
+  return controllers_lists_[free_controllers_list];
+}
+
+const std::vector<ControllerSpec> & ControllerManager::RTControllerListWrapper::get_updated_list(
+  const std::lock_guard<std::recursive_mutex> &) const
```

Comment:
Please check the comment of the function on the header, it's the only way I found to make sure that whoever is calling this has a lock guard and has locked this class mutex (the line below).

While technically they could give me a random lock_guard, and enter this function without having the mutex locked, it would be with malice, this is intended to prevent accidental usage without having locked the mutex.